File: android/Reguerta/app/src/main/java/com/reguerta/user/data/notifications/FirebaseShiftNotificationDetailRepository.kt
package com.reguerta.user.data.notifications

import com.reguerta.user.data.access.AuthenticatedFunctionCaller
import com.reguerta.user.data.access.AuthenticatedFunctionException
import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import com.reguerta.user.domain.notifications.ShiftNotificationDetail
import com.reguerta.user.domain.notifications.ShiftNotificationDetailRepository
import com.reguerta.user.domain.shifts.ShiftAssignment
import com.reguerta.user.domain.shifts.ShiftStatus
import com.reguerta.user.domain.shifts.ShiftType
import java.io.IOException
import kotlinx.coroutines.CancellationException
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put

internal class FirebaseShiftNotificationDetailRepository(
    private val functionCaller: AuthenticatedFunctionCaller,
    private val requestedEnvironment: () -> String,
) : ShiftNotificationDetailRepository {
    override suspend fun getCurrentDetail(
        eventId: String,
        memberId: String,
    ): ShiftNotificationDetail = try {
        val normalizedEventId = eventId.trim().takeIf(String::isNotBlank)
            ?: invalidDetailResponse()
        val normalizedMemberId = memberId.trim().takeIf(String::isNotBlank)
            ?: invalidDetailResponse()
        val response = functionCaller.post(
            functionName = RESOLVE_SHIFT_NOTIFICATION_DETAIL_FUNCTION,
            body = buildJsonObject {
                put("environment", requestedEnvironment())
                put("eventId", normalizedEventId)
            },
        )
        decodeShiftNotificationDetail(
            value = response,
            expectedEventId = normalizedEventId,
            expectedMemberId = normalizedMemberId,
        )
    } catch (error: Exception) {
        throw error.toShiftNotificationDetailException()
    }

    private companion object {
        const val RESOLVE_SHIFT_NOTIFICATION_DETAIL_FUNCTION = "resolveShiftNotificationDetail"
    }
}

internal fun decodeShiftNotificationDetail(
    value: JsonObject,
    expectedEventId: String,
    expectedMemberId: String,
): ShiftNotificationDetail {
    value.requireExactKeys(
        "schemaVersion",
        "eventId",
        "assignmentRevision",
        "documentRevision",
        "shift",
    )
    if (value.requiredLong("schemaVersion") != 1L) invalidDetailResponse()
    val eventId = value.requiredString("eventId")
    if (eventId != expectedEventId) invalidDetailResponse()
    val assignmentRevision = value.requiredPositiveLong("assignmentRevision")
    val documentRevision = value.requiredPositiveLong("documentRevision")
    if (assignmentRevision > documentRevision) invalidDetailResponse()
    val shiftValue = value["shift"] as? JsonObject ?: invalidDetailResponse()
    shiftValue.requireExactKeys(
        "id",
        "type",
        "dateMillis",
        "assignedUserIds",
        "helperUserId",
        "status",
        "source",
        "createdAtMillis",
        "updatedAtMillis",
    )
    val type = when (shiftValue.requiredString("type")) {
        "delivery" -> ShiftType.DELIVERY
        "market" -> ShiftType.MARKET
        else -> invalidDetailResponse()
    }
    val assignedUserIds = shiftValue.requiredStringArray("assignedUserIds")
    val expectedCount = if (type == ShiftType.DELIVERY) 1 else 3
    if (
        assignedUserIds.size != expectedCount ||
        expectedMemberId !in assignedUserIds
    ) {
        invalidDetailResponse()
    }
    val helperUserId = shiftValue.optionalString("helperUserId")
    if ((type == ShiftType.DELIVERY) == (helperUserId == null)) {
        invalidDetailResponse()
    }
    val status = when (shiftValue.requiredString("status")) {
        "planned" -> ShiftStatus.PLANNED
        "swap_pending" -> ShiftStatus.SWAP_PENDING
        "confirmed" -> ShiftStatus.CONFIRMED
        else -> invalidDetailResponse()
    }
    if (shiftValue.requiredString("source") != "app") invalidDetailResponse()
    val createdAtMillis = shiftValue.requiredLong("createdAtMillis")
    val updatedAtMillis = shiftValue.requiredLong("updatedAtMillis")
    if (updatedAtMillis < createdAtMillis) invalidDetailResponse()
    return ShiftNotificationDetail(
        eventId = eventId,
        assignmentRevision = assignmentRevision,
        documentRevision = documentRevision,
        shift = ShiftAssignment(
            id = shiftValue.requiredString("id"),
            type = type,
            dateMillis = shiftValue.requiredLong("dateMillis"),
            assignedUserIds = assignedUserIds,
            helperUserId = helperUserId,
            status = status,
            source = "app",
            createdAtMillis = createdAtMillis,
            updatedAtMillis = updatedAtMillis,
        ),
    )
}

private fun JsonObject.requireExactKeys(vararg expected: String) {
    if (keys != expected.toSet()) invalidDetailResponse()
}

private fun JsonObject.requiredString(key: String): String =
    (this[key] as? JsonPrimitive)?.contentOrNull?.trim()?.takeIf(String::isNotBlank)
        ?: invalidDetailResponse()

private fun JsonObject.optionalString(key: String): String? = when (val value = this[key]) {
    JsonNull -> null
    is JsonPrimitive -> value.contentOrNull?.trim()?.takeIf(String::isNotBlank)
        ?: invalidDetailResponse()
    else -> invalidDetailResponse()
}

private fun JsonObject.requiredLong(key: String): Long =
    (this[key] as? JsonPrimitive)?.longOrNull ?: invalidDetailResponse()

private fun JsonObject.requiredPositiveLong(key: String): Long =
    requiredLong(key).takeIf { it > 0L } ?: invalidDetailResponse()

private fun JsonObject.requiredStringArray(key: String): List<String> =
    (this[key] as? JsonArray)?.map { value ->
        (value as? JsonPrimitive)?.contentOrNull?.trim()?.takeIf(String::isNotBlank)
            ?: invalidDetailResponse()
    } ?: invalidDetailResponse()

private fun invalidDetailResponse(): Nothing = throw RepositoryException(
    kind = RepositoryErrorKind.INVALID_DATA,
    resource = "notifications.shiftDetail.response",
)

private fun Throwable.toShiftNotificationDetailException(): Throwable = when (this) {
    is CancellationException,
    is RepositoryException,
        -> this
    is AuthenticatedFunctionException -> RepositoryException(
        kind = when {
            statusCode == 401 || statusCode == 403 -> RepositoryErrorKind.PERMISSION_DENIED
            statusCode == 404 -> RepositoryErrorKind.NOT_FOUND
            statusCode == 408 || statusCode == 429 || statusCode >= 500 -> RepositoryErrorKind.UNAVAILABLE
            else -> RepositoryErrorKind.UNKNOWN
        },
        resource = "notifications.shiftDetail",
        cause = this,
    )
    is IOException -> RepositoryException(
        kind = RepositoryErrorKind.UNAVAILABLE,
        resource = "notifications.shiftDetail",
        cause = this,
    )
    else -> RepositoryException(
        kind = RepositoryErrorKind.UNKNOWN,
        resource = "notifications.shiftDetail",
        cause = this,
    )
}

File: ios/Reguerta/Reguerta/Data/Notifications/FirebaseShiftNotificationDetailRepository.swift
import Foundation

@MainActor
struct FirebaseShiftNotificationDetailRepository: ShiftNotificationDetailRepository {
    private let storedFunctionsClient: AuthenticatedFirebaseFunctionsClient

    func currentDetail(
        eventID: String,
        memberID: String,
        environment: SessionEnvironment
    ) async throws -> ShiftNotificationDetail {
        let normalizedEventID = eventID.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalizedMemberID = memberID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalizedEventID.isEmpty, !normalizedMemberID.isEmpty else { throw Self.invalidData }

        let response: ShiftNotificationDetailResponse
        do {
            response = try await storedFunctionsClient.post(
                function: .resolveShiftNotificationDetail,
                body: ShiftNotificationDetailRequest(environment: environment, eventId: normalizedEventID),
                response: ShiftNotificationDetailResponse.self
            )
        } catch {
            throw Self.map(error)
        }

        return try response.detail(expectedEventID: normalizedEventID, expectedMemberID: normalizedMemberID)
    }

    nonisolated fileprivate static var invalidData: RepositoryError {
        RepositoryError.invalidData(resource: "notifications.shiftDetail.response")
    }

    private static func map(_ error: any Error) -> any Error {
        if error is CancellationError { return error }
        guard let clientError = error as? FirebaseFunctionClientError else {
            return RepositoryError.unknown(resource: "notifications.shiftDetail")
        }
        switch clientError {
        case .cancelled:
            return CancellationError()
        case .unauthorized, .forbidden, .missingIDToken:
            return RepositoryError.permissionDenied(resource: "notifications.shiftDetail")
        case .http(let statusCode, _, _) where statusCode == 404:
            return RepositoryError.notFound(resource: "notifications.shiftDetail")
        case .timeout, .transport:
            return RepositoryError.unavailable(resource: "notifications.shiftDetail")
        case .http(let statusCode, _, _) where statusCode == 408 || statusCode == 429 || statusCode >= 500:
            return RepositoryError.unavailable(resource: "notifications.shiftDetail")
        case .invalidResponse:
            return invalidData
        case .invalidEndpoint, .conflict, .http, .invalidHTTPResponse:
            return RepositoryError.unknown(resource: "notifications.shiftDetail")
        }
    }
}

extension FirebaseShiftNotificationDetailRepository {
    init(functionsClient: AuthenticatedFirebaseFunctionsClient) {
        storedFunctionsClient = functionsClient
    }
}

nonisolated private struct ShiftNotificationDetailRequest: Encodable {
    let environment: SessionEnvironment
    let eventId: String
}

nonisolated private struct ShiftNotificationDetailResponse {
    let schemaVersion: Int64
    let eventId: String
    let assignmentRevision: Int64
    let documentRevision: Int64
    let shift: ShiftNotificationDetailShiftResponse

    func detail(expectedEventID: String, expectedMemberID: String) throws -> ShiftNotificationDetail {
        guard schemaVersion == 1,
              eventId == expectedEventID,
              assignmentRevision > 0,
              documentRevision > 0,
              assignmentRevision <= documentRevision else {
            throw FirebaseShiftNotificationDetailRepository.invalidData
        }
        return try ShiftNotificationDetail(
            eventID: eventId,
            assignmentRevision: assignmentRevision,
            documentRevision: documentRevision,
            shift: shift.assignment(expectedMemberID: expectedMemberID)
        )
    }
}

extension ShiftNotificationDetailResponse: Decodable {
    init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: ShiftNotificationDetailCodingKey.self)
        try container.requireExactKeys([
            "schemaVersion", "eventId", "assignmentRevision", "documentRevision", "shift"
        ])
        schemaVersion = try container.decode(Int64.self, forKey: .key("schemaVersion"))
        eventId = try container.decode(String.self, forKey: .key("eventId"))
        assignmentRevision = try container.decode(Int64.self, forKey: .key("assignmentRevision"))
        documentRevision = try container.decode(Int64.self, forKey: .key("documentRevision"))
        shift = try container.decode(ShiftNotificationDetailShiftResponse.self, forKey: .key("shift"))
    }
}

nonisolated private struct ShiftNotificationDetailShiftResponse {
    let id: String
    let type: ShiftType
    let dateMillis: Int64
    let assignedUserIds: [String]
    let helperUserId: String?
    let status: ShiftStatus
    let source: String
    let createdAtMillis: Int64
    let updatedAtMillis: Int64

    func assignment(expectedMemberID: String) throws -> ShiftAssignment {
        let normalizedID = id.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalizedAssignedIDs = assignedUserIds.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        let expectedCount = type == .delivery ? 1 : 3
        guard !normalizedID.isEmpty,
              normalizedAssignedIDs.count == expectedCount,
              normalizedAssignedIDs.allSatisfy({ !$0.isEmpty }),
              Set(normalizedAssignedIDs).count == normalizedAssignedIDs.count,
              normalizedAssignedIDs.contains(expectedMemberID),
              source == "app",
              updatedAtMillis >= createdAtMillis else {
            throw FirebaseShiftNotificationDetailRepository.invalidData
        }
        let normalizedHelperID = helperUserId?.trimmingCharacters(in: .whitespacesAndNewlines)
        guard (type == .delivery && normalizedHelperID?.isEmpty == false) ||
                (type == .market && helperUserId == nil) else {
            throw FirebaseShiftNotificationDetailRepository.invalidData
        }
        return ShiftAssignment(
            id: normalizedID,
            type: type,
            dateMillis: dateMillis,
            assignedUserIds: normalizedAssignedIDs,
            helperUserId: normalizedHelperID,
            status: status,
            source: source,
            createdAtMillis: createdAtMillis,
            updatedAtMillis: updatedAtMillis
        )
    }
}

extension ShiftNotificationDetailShiftResponse: Decodable {
    init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: ShiftNotificationDetailCodingKey.self)
        try container.requireExactKeys([
            "id", "type", "dateMillis", "assignedUserIds", "helperUserId", "status", "source",
            "createdAtMillis", "updatedAtMillis"
        ])
        id = try container.decode(String.self, forKey: .key("id"))
        let rawType = try container.decode(String.self, forKey: .key("type"))
        guard let type = ShiftType(rawValue: rawType) else {
            throw FirebaseShiftNotificationDetailRepository.invalidData
        }
        self.type = type
        dateMillis = try container.decode(Int64.self, forKey: .key("dateMillis"))
        assignedUserIds = try container.decode([String].self, forKey: .key("assignedUserIds"))
        helperUserId = try container.decodeIfPresent(String.self, forKey: .key("helperUserId"))
        let rawStatus = try container.decode(String.self, forKey: .key("status"))
        guard let status = ShiftStatus(rawValue: rawStatus) else {
            throw FirebaseShiftNotificationDetailRepository.invalidData
        }
        self.status = status
        source = try container.decode(String.self, forKey: .key("source"))
        createdAtMillis = try container.decode(Int64.self, forKey: .key("createdAtMillis"))
        updatedAtMillis = try container.decode(Int64.self, forKey: .key("updatedAtMillis"))
    }
}

nonisolated private struct ShiftNotificationDetailCodingKey: CodingKey, Hashable {
    let stringValue: String
    let intValue: Int?

    static func key(_ value: String) -> ShiftNotificationDetailCodingKey {
        ShiftNotificationDetailCodingKey(stringValue: value, intValue: nil)
    }
}

extension ShiftNotificationDetailCodingKey {
    init?(stringValue: String) {
        self.stringValue = stringValue
        intValue = nil
    }

    init?(intValue: Int) {
        stringValue = String(intValue)
        self.intValue = intValue
    }

}

nonisolated private extension KeyedDecodingContainer where Key == ShiftNotificationDetailCodingKey {
    func requireExactKeys(_ expected: Set<String>) throws {
        guard Set(allKeys.map(\.stringValue)) == expected else {
            throw FirebaseShiftNotificationDetailRepository.invalidData
        }
    }
}
